### PR TITLE
Created a minibatch iterator selection function

### PR DESCRIPTION
**Pull request by farhadrclass**
as discussed here https://github.com/JuliaSmoothOptimizers/KnetNLPModels.jl/pull/52

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -44,20 +44,76 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+  rand_minibatch_train!(nlp::AbstractKnetNLPModel)
+
+Select a training minibatch for `nlp` randomly.
+"""
+rand_minibatch_train!(nlp::AbstractKnetNLPModel) =
   nlp.current_training_minibatch = rand(nlp.training_minibatch_iterator)
 
+"""
+minibatch_next_train!(nlp::AbstractKnetNLPModel)
+Selects the next mini-batch from training_minibatch_iterator
```
Comment (paraynaud) [inline src/utils.jl:62]:
```suggestion
    minibatch_next_train!(nlp::AbstractKnetNLPModel)

Selects the next mini-batch from training_minibatch_iterator
```

File: src/utils.jl [hunk 2]
```diff
@@ -44,20 +44,76 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+nlp:: KnetNLPModel 
+i:: current location in the iterator
+Returns a new current location of the iterator i, if 0 is returned when all mini-batches are visited
+can be used in a loop or method call - refere to KnetNLPModelProblems for more use cases
+"""
+function minibatch_next_train!(nlp::AbstractKnetNLPModel, i::int)
+  i += nlp.size_minibatch # update the i by mini_batch size
```
Comment (paraynaud) [inline src/utils.jl:69]: `nlp.size_minibatch` is an Integer telling in how many pieces the training dataset is divided.
I am not sure about the effect of this line.
Maybe you want something similar to 
```julia
i += length(nlp.training_data_set / nlp.size_minibatch 
```

Maybe you should add properly `i_training` and `i_test` in the `KnetNLPModel` structure.
This way you could have 
```julia
function minibatch_next_train!(nlp::AbstractKnetNLPModel; i::int=nlp.i)
```

File: src/utils.jl [hunk 3]
```diff
@@ -44,20 +44,76 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
 """
     reset_minibatch_train!(nlp::AbstractKnetNLPModel)
 
-Select a new training minibatch for `nlp`.
-Typically used before a new evaluation of the loss function/gradient.
+Select a the first training minibatch for `nlp`.
 """
 reset_minibatch_train!(nlp::AbstractKnetNLPModel) =
+  nlp.current_training_minibatch = first(nlp.training_minibatch_iterator)
```
Comment (paraynaud) [inline src/utils.jl:50]: I think your method `minibatch_next_train!` should be the default `reset_minibtach_train!`.
You may add a function `first_minibatch_train!`, that you can call it in 
`minibatch_next_train!` when ` i >= nlp.training_minibatch_iterator.imax `.

File: src/utils.jl [hunk 4]
```diff
@@ -44,20 +44,76 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+  i += nlp.size_minibatch #TODO in the futue we might want to have different size for minbatch test vs train
+  if (i >= nlp.test_minibatch_iterator.imax)
+    nlp.current_test_minibatch = first(nlp.test_minibatch_iterator)
+    return 0
+  else
+    next = iterate(nlp.test_minibatch_iterator, i)
+    nlp.current_test_minibatch = next[1]
+    return i
+  end
+end
```
Comment (paraynaud) [inline src/utils.jl:115]: Same comments.

File: src/utils.jl [hunk 5]
```diff
@@ -44,20 +44,76 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+"""
+function minibatch_next_train!(nlp::AbstractKnetNLPModel, i::int)
+  i += nlp.size_minibatch # update the i by mini_batch size
+  if (i >= nlp.training_minibatch_iterator.imax)
+    # reset to the begining and return zero 
+    nlp.current_training_minibatch = first(nlp.training_minibatch_iterator) # reset to the first one
+    return 0
+  else
+    next = iterate(nlp.training_minibatch_iterator, i)
+    nlp.current_training_minibatch = next[1]
```
Comment (paraynaud) [inline src/utils.jl:76]:
```suggestion
    (next, indice) = iterate(nlp.training_minibatch_iterator, i)
    nlp.current_training_minibatch = next
```

**Review by paraynaud — CHANGES_REQUESTED** [review]
Thanks, and don't forget to add some tests!

File: src/utils.jl (hunk 3 above)
Comment (farhadrclass) [reply src/utils.jl:50]: but reset usually mean resetting to the first occurrences or default set, that is why I added that as that

File: src/utils.jl (hunk 2 above)
Comment (farhadrclass) [reply src/utils.jl:70]: I agree with  i_training and i_test should be there  in the KnetNLPModel structure I will do the modification 

as a matter of fact I checked the  i += length(nlp.training_data_set / nlp.size_minibatch ) doesn't work 
The way knet does it is to ignore the size and do {i \in [1,data_size-batch_size]} which is terrible but that mean i for batch size of 1000 in MNIST is between 1 to 59999 which is weird for me but that is how it was written (I spent 2 days debugging till I find this weird behavior )

File: src/utils.jl (hunk 4 above)
Comment (farhadrclass) [reply src/utils.jl:116]: Same issue as I mentioned earlier

**Comment by farhadrclass** [comment]
I will re-factor the code and also add the tests now

File: src/utils.jl (hunk 2 above)
Comment (farhadrclass) [reply src/utils.jl:70]: check here the knet random code: https://github.com/denizyuret/Knet.jl/blob/e9bec84b826ee7231db4ca8772fae6eb2fd4bcb9/src/train20/data.jl

function rand(d::Data)
    i = rand(0:(d.length-d.batchsize))
    return iterate(d, i)[1]
end

Comment (paraynaud) [reply src/utils.jl]: Did you try `length(nlp.training_data_set) / nlp.size_minibatch`, I forgot the coma in my previous answer.

```quote
that mean i for batch size of 1000 in MNIST is between 1 to 59999
```
I think the important point is: i <= training_minibatch_iterator.imax

File: src/utils.jl (hunk 3 above)
Comment (paraynaud) [reply src/utils.jl]: Hmm... I will have to update my code then --'.

File: Project.toml [hunk 6]
```diff
@@ -5,6 +5,7 @@ version = "0.2.1"
 
 [deps]
 Knet = "1902f260-5fb4-5aff-8c31-6271790ab950"
+MLDatasets = "eb30cadb-4394-5ae3-aed4-317e484a6458"
```
Comment (paraynaud) [inline Project.toml:8]: MLDatasets must not be a dependency.

```suggestion
```

File: src/utils.jl [hunk 7]
```diff
@@ -44,19 +44,81 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+  nlp.i = rand(1:(nlp.training_minibatch_iterator.imax))
+  nlp.current_training_minibatch = iterate(nlp.training_minibatch_iterator, nlp.i)
+end
+
+"""
+    minibatch_next_train!(nlp::AbstractKnetNLPModel)
+
+Selects the next mini-batch from training_minibatch_iterator
+nlp:: KnetNLPModel 
+i:: current location in the iterator
```
[the remark this answers is not in the document]
Comment (farhadrclass) [reply src/utils.jl:69]: good catch, I need to update this

File: src/utils.jl [hunk 8]
```diff
@@ -44,19 +44,81 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+"""
+function rand_minibatch_test!(nlp::AbstractKnetNLPModel)
+  nlp.i_test = rand(1:(nlp.test_minibatch_iterator.imax))
+  nlp.current_test_minibatch = iterate(nlp.test_minibatch_iterator, nlp.i_test)
+end
+
 """
-reset_minibatch_test!(nlp::AbstractKnetNLPModel) =
-  nlp.current_test_minibatch = rand(nlp.test_minibatch_iterator)
+    reset_minibatch_train!(nlp::AbstractKnetNLPModel)
```
[the remark this answers is not in the document]
Comment (farhadrclass) [reply src/utils.jl:96]: oh the name in the doc string

**Comment by farhadrclass** [comment]
@paraynaud  for some reason the doc builder fails 
since it needs knetnlpmodels to run the example and we haven't push the PR, it doesn't have the functions we just wrote

we have to option
accept the PR and it should fix this or make a new one and remove the tutorial to that one

File: docs/src/tutorial.md [hunk 9]
```diff
@@ -103,10 +103,20 @@ That way, the accuracy will not fluctuate with the minibatch.
 
 ## Default behavior
 By default, the training minibatch that evaluates the neural network doesn't change between evaluations.
-To change the training minibatch, use:
+To change the training minibatch, use one of the following methods:
+* To select randomly a mini-batch 
+```@example KnetNLPModel
+rand_minibatch_train!!(DenseNetNLPModel)
```
Comment (paraynaud) [inline docs/src/tutorial.md:109]:
```suggestion
rand_minibatch_train!(DenseNetNLPModel)
```
I think `!!` makes the documentation fails inevitably.

Comment (paraynaud) [reply docs/src/tutorial.md:109]: I let you commit the suggestion, because I don't want to mess your local branch if you made changes since your last git push.

Comment (farhadrclass) [reply docs/src/tutorial.md:109]: oh, nice catch

File: src/utils.jl [hunk 10]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
 """
     reset_minibatch_train!(nlp::AbstractKnetNLPModel)
 
-Select a new training minibatch for `nlp`.
-Typically used before a new evaluation of the loss function/gradient.
+Select a the first training minibatch for `nlp`.
```
Comment (paraynaud) [inline src/utils.jl:47]:
```suggestion
Select the first training minibatch for `nlp`.
```

File: src/utils.jl [hunk 11]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
-  nlp.current_test_minibatch = rand(nlp.test_minibatch_iterator)
+function rand_minibatch_test!(nlp::AbstractKnetNLPModel)
+  nlp.i_test = rand(1:(nlp.test_minibatch_iterator.imax))
+  nlp.current_test_minibatch = iterate(nlp.test_minibatch_iterator, nlp.i_test)
+end
+
+"""
+    reset_minibatch_test!(nlp::AbstractKnetNLPModel)
+
+Select a the first test minibatch for `nlp`.
```
Comment (paraynaud) [inline src/utils.jl:97]:
```suggestion
Select the first test minibatch for `nlp`.
```

File: src/utils.jl [hunk 12]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+  nlp.i_train = rand(1:(nlp.training_minibatch_iterator.imax))
+  nlp.current_training_minibatch = iterate(nlp.training_minibatch_iterator, nlp.i_train)
+end
+
+"""
+    minibatch_next_train!(nlp::AbstractKnetNLPModel)
+
+Selects the next mini-batch from `nlp.training_minibatch_iterator`.  
+Returns a new current location of the iterator i, if 1 is returned when all mini-batches are visited
+can be used in a loop or method call - refere to KnetNLPModelProblems for more use cases
```
Comment (paraynaud) [inline src/utils.jl:69]:
```suggestion
Returns the new current location of the iterator `nlp.i_train`.
If it returns 1, the current training minibatch is the first of `nlp.training_minibatch_iterator` and the previous minibatch was the last of `nlp.training_minibatch_iterator`.
`minibatch_next_train!` aims to be used in a loop or method call - refere to KnetNLPModelProblems.jl for more use cases
```

File: src/utils.jl [hunk 13]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+end
+
+"""
+    minibatch_next_test!(nlp::AbstractKnetNLPModel)
+
+Selects the next mini-batch from test_minibatch_iterator
+nlp:: KnetNLPModel 
+i:: current location in the iterator
+Returns a new current location of the iterator i, if 1 is returned when all mini-batches are visited
+can be used in a loop or method call - refere to KnetNLPModelProblems for more use cases
```
Comment (paraynaud) [inline src/utils.jl:111]:
```suggestion
Selects the next mini-batch from `test_minibatch_iterator`.
Returns the new current location of the iterator `nlp.i_test`.
If it returns 1, the current training minibatch is the first of `nlp.test_minibatch_iterator` and the previous minibatch was the last of `nlp.test_minibatch_iterator`.
`minibatch_next_test!` aims to be used in a loop or method call - refere to KnetNLPModelProblems.jl for more use cases
```

**Review by paraynaud — COMMENTED** [review]
I checked the documentation of the new methods.
I made some changes to ease the understanding.
If you agree with the new doc, we will ask Dominique to finally review the PR.

**Comment by paraynaud** [comment]
@dpo, this one should be ready to go.

File: docs/src/tutorial.md [hunk 14]
```diff
@@ -103,10 +103,20 @@ That way, the accuracy will not fluctuate with the minibatch.
 
 ## Default behavior
 By default, the training minibatch that evaluates the neural network doesn't change between evaluations.
-To change the training minibatch, use:
+To change the training minibatch, use one of the following methods:
+* To select randomly a mini-batch 
+```@example KnetNLPModel
+rand_minibatch_train!(DenseNetNLPModel)
+```
+* To select the next mini-batch from current mini-batch iterator (Can be used in a loop to go over all the datasets)
```
Comment (dpo) [inline docs/src/tutorial.md:111]:
```suggestion
* To select the next mini-batch from current mini-batch iterator (Can be used in a loop to go over the entire dataset)
```

File: docs/src/tutorial.md [hunk 15]
```diff
@@ -103,10 +103,20 @@ That way, the accuracy will not fluctuate with the minibatch.
...
+```@example KnetNLPModel
+minibatch_next_train!(DenseNetNLPModel)
+```
+* To reset the first mini-batch
```
Comment (dpo) [inline docs/src/tutorial.md:115]: What do you mean here? Reset *to* the first mini batch?

File: src/KnetNLPModels.jl [hunk 16]
```diff
@@ -99,6 +103,8 @@ function KnetNLPModel(
     x0,
     layers_g,
     nested_array,
+    1, #inti the batch current i to 1
```
Comment (dpo) [inline src/KnetNLPModels.jl:106]: I don't understand the comment. Index of the first minibatch? Normally, you don't index iterators. You just loop through them.

File: src/utils.jl [hunk 17]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
 """
     reset_minibatch_train!(nlp::AbstractKnetNLPModel)
 
-Select a new training minibatch for `nlp`.
-Typically used before a new evaluation of the loss function/gradient.
+Select the first training minibatch for `nlp`.
```
Comment (dpo) [inline src/utils.jl:47]: Spell "mini batch" the same way everywhere.

File: src/utils.jl [hunk 18]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
 """
-reset_minibatch_train!(nlp::AbstractKnetNLPModel) =
-  nlp.current_training_minibatch = rand(nlp.training_minibatch_iterator)
+function reset_minibatch_train!(nlp::AbstractKnetNLPModel)
+  nlp.current_training_minibatch = first(nlp.training_minibatch_iterator)
+  nlp.i_train = 1
```
Comment (dpo) [inline src/utils.jl:51]: Here, we see that the index 1 is not useful.

File: src/utils.jl [hunk 19]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+end
+
+"""
+  rand_minibatch_train!(nlp::AbstractKnetNLPModel)
+
+Select a training minibatch for `nlp` randomly.
+"""
+function rand_minibatch_train!(nlp::AbstractKnetNLPModel)
+  nlp.i_train = rand(1:(nlp.training_minibatch_iterator.imax))
+  nlp.current_training_minibatch = iterate(nlp.training_minibatch_iterator, nlp.i_train)
```
Comment (dpo) [inline src/utils.jl:61]: What is the point of this method? Why not just grab the next mini batch in the iterator?

File: src/utils.jl [hunk 20]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+end
+
+"""
+    minibatch_next_train!(nlp::AbstractKnetNLPModel)
+
+Selects the next mini-batch from `nlp.training_minibatch_iterator`.  
+Returns the new current location of the iterator `nlp.i_train`.
+If it returns 1, the current training minibatch is the first of `nlp.training_minibatch_iterator` and the previous minibatch was the last of `nlp.training_minibatch_iterator`.
+`minibatch_next_train!` aims to be used in a loop or method call - refere to KnetNLPModelProblems.jl for more use cases
```
Comment (dpo) [inline src/utils.jl:70]:
```suggestion
`minibatch_next_train!` aims to be used in a loop or method call.
Refer to KnetNLPModelProblems.jl for more use cases.
```

File: src/utils.jl [hunk 21]
```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
...
+"""
+function minibatch_next_train!(nlp::AbstractKnetNLPModel)
+  nlp.i_train += nlp.size_minibatch # update the i by mini_batch size
+  if (nlp.i_train >= nlp.training_minibatch_iterator.imax)
+    # reset to the begining 
+    nlp.current_training_minibatch = first(nlp.training_minibatch_iterator) # reset to the first one
+    nlp.i_train = 1
+  else
```
Comment (dpo) [inline src/utils.jl:78]: why not use `try ... catch` to reset to `first` when `iterate` returns `nothing`?

File: src/utils.jl (hunk 18 above)
Comment (farhadrclass) [reply src/utils.jl:51]: we use the index 1 as a flag for when data is visited or new iteration has started

File: src/KnetNLPModels.jl (hunk 16 above)
Comment (farhadrclass) [reply src/KnetNLPModels.jl:106]: yes but this is because Knet doesn't use the iterator correctly 

The way knet does it is to ignore the size and do {i \in [1,data_size-batch_size]} which is terrible but that mean i for batch size of 1000 in MNIST is between 1 to 59999 which is weird for me but that is how it was written (I spent 2 days debugging till I find this weird behavior )


check here the knet random code: https://github.com/denizyuret/Knet.jl/blob/e9bec84b826ee7231db4ca8772fae6eb2fd4bcb9/src/train20/data.jl

function rand(d::Data)
i = rand(0:(d.length-d.batchsize))
return iterate(d, i)[1]
end

So basically their iterator is using i and i actually is the index of the data not the mini-batch, 
I am open to suggestion how to fix this since I think the problem is Knet's iterator is to some extend weird

File: src/utils.jl (hunk 17 above)
Comment (farhadrclass) [reply src/utils.jl:47]: yes will update them

File: src/utils.jl (hunk 19 above)
Comment (farhadrclass) [reply src/utils.jl:61]: This is for selecting a random batch, 
as some of the DNN training, instead of all data, they select a minibatch as random, this way we support that as well

Comment (paraynaud) [reply src/utils.jl:61]: The PR goal is to provide several choices to select a training mini-batch from the training dataset.
This method is the current `reset_minibatch_train!` from main renamed, we didn't want to make KnetNLPModel functionnalities regress.
We changed its name for consistency.

File: src/utils.jl (hunk 21 above)
Comment (farhadrclass) [reply src/utils.jl:80]: I suggest we put this in a if else since the iterator function returns nothing if out of bound.
I think if else is less memory heavy than catch-try, no?
but I changed the code to be more robust

Comment (paraynaud) [reply src/utils.jl:80]: @dpo I understoof from my course that Exceptions must be avoid, to not slow down the execution with system calls?
It is implemented differently in julia?

Comment (dpo) [reply src/utils.jl:80]: Actually you wouldn't need a `try ... catch` because `iterate` returns `nothing` (it doesn't raise an exception). It seems strange to me to index an iterator.

Comment (farhadrclass) [reply src/utils.jl:80]: exactly what I thought first I used it 
but Knet uses this code for iterate: https://github.com/denizyuret/Knet.jl/blob/master/src/train20/data.jl not sure if this is correct or not 
Here is their iterated:
`
@propagate_inbounds function iterate(d::Data, i=0)     # returns data in d.indices[i+1:i+batchsize]
    if i >= d.imax
        return nothing
    end
    if d.shuffle && i == 0
        d.indices = randperm(d.length)
    end
    nexti = min(i + d.batchsize, d.length)
    ids = d.indices[i+1:nexti]
    xbatch = try convert(d.xtype, reshape(d.x[:,ids],d.xsize[1:end-1]...,length(ids)))
    catch e; throw(DimensionMismatch("Error converting batch from $(summary(d.x)) to $(d.xtype): $e")); end
    if d.y == nothing
        return (xbatch,nexti)
    else
        ybatch = try convert(d.ytype, reshape(d.y[:,ids],d.ysize[1:end-1]...,length(ids)))
        catch; throw(DimensionMismatch("Y tensor not compatible with size=$(d.ysize) and type=$(d.ytype)")); end
        return ((xbatch,ybatch),nexti)
    end
end


@propagate_inbounds function iterate(r::Repeat, s=(1,))
    epoch, state = s[1], tail(s)
    epoch > r.n && return nothing
    next = iterate(r.data, state...)
    next === nothing && return iterate(r, (epoch+1,))
    (next[1], (epoch, next[2]))
end


`

Comment (farhadrclass) [reply src/utils.jl:79]: @dpo  I updated the code or put comments that needs your input, let me know what you think

**Review by dpo — DISMISSED** [review]
Thanks! Just a few minor comments. You can merge after that.

File: src/KnetNLPModels.jl (hunk 16 above)
Comment (paraynaud) [inline src/KnetNLPModels.jl:106]:
```suggestion
    1, #initialize the batch current i to 1
```

**Review by paraynaud — APPROVED** [review]
